Forms/update responses management to use DataViews

ntsekouras opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Resolves https://github.com/Automattic/jetpack/issues/41122

Currently there are two available UIs for managing the responses. One is a wp-admin list and the other is a JS custom list view.

The goal is to replace the existing two views with a single one powered by DataViews and focus on a single core-first modern UI, without having to reimplement filtering, sorting mechanisms, etc..

This PR for now updates the JS view with the help of DataViews.

### Video (as of 6th March)


https://github.com/user-attachments/assets/180d41d2-2b9e-4707-a8fc-2e929f176df1




### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?


## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
1. Go to `Feedback` screen (`/wp-admin/admin.php?page=jetpack-forms`)
2. You need to have at least one response at your site
3. Make sure you select the `Inbox` view from the `View` dropdown menu at the top right corner of the screen.
4. Play around with the flows, filters, etc..

ntsekouras commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -9,12 +9,441 @@
+	 * @param array $schema Schema array.
+	 * @return array Modified Schema array.
+	 */
+	public function add_additional_fields_schema( $schema ) {
```
@anton-vlasenko the PHP changes are the first pass, but is there a better way to do this? 

We want to add more fields to the response object. It's also related to the `prepare_item_for_response` [below](https://github.com/Automattic/jetpack/pull/41602/files#diff-9db38de6e3678fb163907d4d47e3b1f38a9a74e038e9b5b9f475c190c0e8a937R176).

anton-vlasenko replied:
I wonder if you could use [something along the lines of](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-attachments-controller.php#L942) `WP_REST_Attachments_Controller::get_item_schema()`?
Since the `Contact_Form_Endpoint` class extends `WP_REST_Posts_Controller`, you could call `parent::get_item_schema()` and then add your own fields to the schema - just as `WP_REST_Attachments_Controller::get_item_schema()` does. here.
What do you think?

anton-vlasenko commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -9,12 +9,441 @@
+		$post_ids = $request->get_param( 'post_ids' );
+
+		if ( $action && ! is_array( $post_ids ) ) {
+			return new $this->error_response( __( 'Bad request', 'jetpack-forms' ), 400 );
```
```suggestion
			return $this->error_response( __( 'Bad request', 'jetpack-forms' ), 400 );
```

This code calls a constructor on an object that has already been instantiated by error_response. This should result in a fatal error: https://3v4l.org/kKJYT.

ntsekouras replied:
Good stuff here. Thanks! I've tried a few times to commit your suggestions through GH and it has some problems. I'll do it manually.

Also noting that especially the `bulk actions` code was just copied over and it will be good to check again to see what we can improve. 

One simple example is to see if we can remove some of the functions and handle some updates through REST API by updating the status etc..

anton-vlasenko commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -9,12 +9,441 @@
+				'date'   => array_map(
+					function ( $row ) {
+						return array(
+							'month' => intval( $row->month ),
+							'year'  => intval( $row->year ),
```
Nitpick: `(int)` is faster than `intval()`.
```suggestion
							'month' => (int) $row->month,
							'year'  => (int) $row->year,
```

anton-vlasenko commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -9,12 +9,441 @@
+					$months
+				),
+				'source' => array_map(
+					function ( $post_id ) {
```
```suggestion
					static function ( $post_id ) {
```

anton-vlasenko commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -9,12 +9,441 @@
+		return rest_ensure_response(
+			array(
+				'date'   => array_map(
+					function ( $row ) {
```
```suggestion
					static function ( $row ) {
```

anton-vlasenko reviewed (commented):
Here's a quick review of the PR (I understand it's still in draft). Thank you.

ntsekouras commented on .pnpmfile.cjs:
```diff
@@ -41,6 +41,36 @@ function fixDeps( pkg ) {
+	// https://github.com/WordPress/gutenberg/issues/67864
+	if ( pkg.name === '@wordpress/dataviews' ) {
+		for ( const dep of [
+			'change-case',
```
How did you compile this list? Was it by looking at the errors during building?

ntsekouras commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -9,12 +9,406 @@
+		$schema = parent::get_item_schema();
+
+		$schema['properties']['uid'] = array(
+			'description' => __( 'The UID... Updated description.', 'jetpack-forms' ),
```
@Automattic/zap I was wondering if you could help with some of these missing descriptions.

lezama replied:
Will help you with this first thing Tomorrow

lezama replied:
[pushed improved descriptions](https://github.com/Automattic/jetpack/pull/41602/commits/e41f64dcfd94d4962580a235089f704bf1a28d65)

ntsekouras commented on projects/packages/forms/src/dashboard/inbox/dataviews/actions/index.js:
```diff
@@ -0,0 +1,223 @@
+	label: __( 'Mark as spam', 'jetpack-forms' ),
+	isEligible: item => item.status !== 'spam',
+	supportsBulk: true,
+	icon: <Icon icon={ spam } />,
```
I had to do this in order to get the proper height/width in the `icon` (24). I'm not sure why this is happening though, as I didn't observe any differences from the `icons` in site editor lists with DataViews, regarding their consumption.

The difference though is that in the other lists probably [this path is returned](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/icon/index.tsx#L95), while here it didn't.

@ciampo @tyxla any thoughts about this?

tyxla replied:
I'm not sure I understand, @ntsekouras, can you please elaborate on what the issue is?

ntsekouras replied:
Sure. If you just add `icon: spam`, as core does, and not the `Icon` wrapper, the icons are bigger than the other lists in core. I couldn't understand why this is happening.

In core the height/width are smaller because the SVG gets the `height/width` attributes, probably by this [code path](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/icon/index.tsx#L95) and in this PR here it gets the `size` prop.

Makes sense?

tyxla replied:
Can't we just pass a `size` prop? I've seen this done for both `Icon` and `Gridicon` quite extensively.

ntsekouras replied:
Ideally, we wouldn't need the Icon wrapper at all, like in core.

tyxla replied:
The core package [literally suggests](https://github.com/WordPress/gutenberg/tree/trunk/packages/icons#usage) using the `Icon` wrapper though:

```
import { Icon, check } from '@wordpress/icons';

<Icon icon={ check } />;
```

maybe I'm missing something?

ciampo replied:
If I understand correctly, dataviews internally passes the `icon` to a `@wordpress/components` `Button` via its `icon` prop: https://github.com/WordPress/gutenberg/blob/75ef79ec8dc0ecefe4c3b180ca69d39a5445d02e/packages/dataviews/src/components/dataviews-item-actions/index.tsx#L77

In that case, I guess the `<Icon />` wrapper wouldn't be necessary, since the `Button` should be able to handle icons from the `@wordpress/icons` package directly.

If that's the case, maybe the answer is in how the `spam` icon is defined, compared to how `@wordpress/icons` are defined (especially in terms of viewbox and explicit width/height)

ntsekouras replied:
Actually it's not only about the `spam` (introduced here), but other icons imported from `@wordpress/icons`.

tyxla replied:
It's also fair to say that `Button` has both `icon` and `iconSize`, so the DataViews actions API might make sense to support `iconSize` as well, if we have a reasonable need for that.

ciampo replied:
> Actually it's not only about the `spam` (introduced here), but other icons imported from `@wordpress/icons`.

We should investigate and understand what's the matter. My assumption is that they would play nicely out of the box.

> It's also fair to say that `Button` has both `icon` and `iconSize`, so the DataViews actions API might make sense to support `iconSize` as well, if we have a reasonable need for that.

My preference would be to keep the API layer as simple as possible.

Also, I'm not a fan of `Button`'s API approach, as it leads to an explosion of props. A better approach would be to allow for a render prop, so that the consumer can pass the `<Icon />` component and specify any other prop directly from consumer-land.

lezama commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -9,12 +9,406 @@
+			$data['entry_permalink'] = $all_fields['entry_permalink'];
+		}
+		if ( rest_is_field_included( 'subject', $fields ) ) {
+			$data['subject'] = $data['_feedback_subject'];
```
I am seeing this warning in the logs:

`[26-Feb-2025 12:20:57 UTC] PHP Warning:  Undefined array key "_feedback_subject" in /usr/local/src/jetpack-monorepo/projects/packages/forms/src/contact-form/class-contact-form-endpoint.php on line 284`

ntsekouras replied:
Thanks for pushing the fix!

lezama replied:
https://github.com/Automattic/jetpack/pull/41602/commits/86f4c439620368d97584b1cc4a10849e7378289d

jasmussen commented:
Hey, took a look at seeing if we could replace the min-height values with flex values. We can, but due to the very deep nesting levels, there are multiple levels to it, and I wasn't able to find all the places in the code where this was written, so instead I wrote this CSS separately, which if applied to the page, works as intended:

![2025-02-26 14 06 13 localhost ec535a8c08a0](https://github.com/user-attachments/assets/897f5900-87df-47ac-8396-fa96cd2cc21c)

Here's the CSS I wrote, annotated:

```
/* We need to make the available canvas 100% tall. Without this, no flex values will work, they will only use the available space. */
.toplevel_page_jetpack-forms #wpwrap,
.toplevel_page_jetpack-forms #wpcontent,
.toplevel_page_jetpack-forms #wpbody,
.toplevel_page_jetpack-forms #wpbody-content {
	height: 100%;
}

#jp-forms-dashboard,
.jp-forms__table,
.jp-forms__layout {
	height: 100%;
}

/* This was just for debugging. Remove these min-heights. */
#jp-forms-dashboard,
.jp-forms__table {
	min-height: 0 !important;
}

/* Now we can begin flexing. */
.jp-forms__layout {
	display: flex;
	flex-direction: column;
}

.jp-forms__logo-wrapper,
.jp-forms__layout-header {
	flex: 0 1 auto; /* Flex, but don't grow */
}

.jp-forms__inbox-tabs {
	flex: 1 1 auto; /* Flex, and grow */

	/* Continue flexing downwards from here. */
	display: flex;
	flex-direction: column;
}

.jp-forms__inbox-tabs .components-tab-panel__tabs {
	flex: 0 1 auto;
}

.jp-forms__inbox-tabs .components-tab-panel__tab-content {
	flex: 1 1 auto;
	
	display: flex;
	flex-direction: column;
}


.jp-forms__inbox-actions {
	flex: 0 1 auto;
}

.jp-forms__inbox-content {
	flex: 1 1 auto;
}
```

The layout is pretty simple. The problems here is that:

* The available canvas wasn't 100% height.
* There are multiple nested flex grids all with virtually the same basic layout: shrink, shrink, grow.

That generic layout can be described as **.shrink**, and **.grow**. Consider then this setup:

```
<html>
<body>
<div class="canvas">

	<h1 class="shrink">Title</h1>
	<section class="grow">Grow</section>

</div>
</body>
</html>
```

In the above reduced setup, everything works as we want it to, provided that:

* html, body, and .canvas are 100% tall
* .canvas is display: flex; and flex-direction: column;
* .shrink is flex: 0 1 auto; (which allows it to shrink but has an automatic height)
* .grow is flex: 1 1 auto; (which allows it to grow, and uses available space)

Hope that helps!

ntsekouras committed:
Forms/update responses management

ntsekouras commented:
@jasmussen thanks so much, but unfortunately you have updated the old view and not the one with DataViews..

jasmussen commented:
> but unfortunately you have updated the old view and not the one with DataViews..

Ack, I guess I didn't succeed in compiling your branch after all!

Can you use the principles I extracted, however? I imagine the principles will be largely the same.

ntsekouras commented:
Thank you for testing @enejb!

>On the desktop. View if you click on the ( eye icon or the view response action) you end up in this state. 

I'll add the `view action` only on mobile.

>On Mobile when you click to view the View Entry. You can't go back via the back swipe or back button to get back to the list view.

I'd say that's expected since it's a Modal, no?

>On Mobile. Inside the mobile view there is both horizontal and vertical scroll inside the data view component. The horizontal scroll can be fixed by the user Once they remove enough columns. But there is still a double vertical scroll.

I'll see what tweaks we can make there, but tables are a bit harder on mobile in general.


>On Mobile. If you click on Export button you end up in a situation where you can't dismiss the export modal.

That's on trunk too actually, with the difference that it doesn't render the button on mobile. Also it uses a different Modal and not the core one, so maybe we should update it to use the core one too. Either way, we should address this.


>Can you use the principles I extracted, however? I imagine the principles will be largely the same.

@jasmussen I'll try again soon. Although  I tried similar things before without success. Let's see how this goes 😄 .

ciampo commented on projects/packages/forms/src/dashboard/inbox/dataviews/index.js:
```diff
@@ -0,0 +1,289 @@
+		],
+		[ filterOptions ]
+	);
+	const actions = useMemo( () => {
+		return [
+			viewAction,
+			markAsSpamAction,
+			markAsNotSpamAction,
+			moveToTrashAction,
+			restoreAction,
+			deleteAction,
+		];
+	}, [] );
```
Nit: can we define this array outside of the component render function, since it looks like a constant value?

ntsekouras replied:
In my next commit it will be not. Actually I was expecting possible feedback to change actions on mobile :)

ntsekouras committed:
address some feedback

jasmussen commented:
> I'll try again soon. Although I tried similar things before without success. Let's see how this goes 😄 .

Sounds good, otherwise let me try again. I have to find out this build system at some point anyway. Though I could swear I followed your steps correctly, and chose the new inbox view in the "View" dropdown.

ntsekouras committed:
use core Modal for export - which fixes mobile view

ntsekouras committed:
more cleanup

ntsekouras committed:
make changes to DataViews calculated height

ntsekouras commented:
@jasmussen I gave it a go once more [here](https://github.com/Automattic/jetpack/pull/41602/commits/e554220d799626732e67447ec4b1e89aa1471c97), but I couldn't find a css only solution, so I've added some JS too to dynamically get the layout header height. It still uses hardcoded values for tabs and the logo, which I expect not to change height in smaller viewports and I tried to avoid more complexity.

It's very fragile though because it will need update if we add more things in the future above the list - which I believe we shouldn't though.

I've also added a css `min-height` in case of quite small in height but with big width viewports.

Maybe it could be fine for first iteration..

ntsekouras committed:
remove side panel transition

jasmussen committed:
Remove JS based height, use flex.

edanzer commented on projects/packages/forms/src/dashboard/inbox/dataviews/index.js:
```diff
@@ -0,0 +1,297 @@
+				return _serachParams;
+			} );
+		},
+		[ data, setSearchParams, isMobile ]
```
Small typo on "_serachParams".

edanzer commented on projects/packages/forms/src/dashboard/inbox/export-responses/google-drive.js:
```diff
@@ -115,15 +113,16 @@ const GoogleDriveExport = ( { onExport } ) => {
 						>
 							{ __( 'Connect to Google Drive', 'jetpack-forms' ) }
-						</a>
+						</Button>
```
There are some minor styling issues with the Connect to Google Drive button. Can probably be solved in a follow up. 

<img width="874" alt="google-button-styles" src="https://github.com/user-attachments/assets/b3db2c4c-8b9b-4e70-aed1-f706c8aa7e2e" />

edanzer commented on projects/packages/forms/src/dashboard/inbox/export-responses/index.js:
```diff
@@ -0,0 +1,83 @@
+		const { getSelectedResponsesFromCurrentDataset, getCurrentQuery } = select( STORE_NAME );
+		return { selected: getSelectedResponsesFromCurrentDataset(), currentQuery: getCurrentQuery() };
+	}, [] );
+	const openModal = useCallback( () => setShowExportModal( true ), [ setShowExportModal ] );
```
Across the PR, it seems like we're using useCallback whenever a function is passed to a component. I wonder if most of these uses are having much impact?

In this case, openModal is passed to a button with no children. Whether we re-render the button component each time or not probably has negligible performance impact.

More broadly, my understanding is that useCallback won't have an impact at all if the child component is not memoized. In this case, if ExportResponses is re-rendered, the button will be re-rendered, and wrapping openModal in useCallback won't prevent that. You'd have to memoize the button. 

Not a huge deal, but my preference would be to limit useCallback to situations where it has a notable impact.

ntsekouras replied:
I agree with you and you're right about not being impactful, but there's an `eslint` rule for that right now and that's why I'm using the `useCallback` in more cases than I'd normally would.

ntsekouras committed:
remove layout header ref

ntsekouras committed:
fix typo

ntsekouras committed:
regenerate pnpm-lock.yaml

ntsekouras committed:
cleanup + keep queryArgs in local state

ntsekouras committed:
fix landing styles + add text domain to translation functions aliases

ntsekouras committed:
remove phan reference to deleted file + keep a sinle `export` label

ntsekouras committed:
fix typo

ntsekouras committed:
refactor DataViews actions and parts of the store

ntsekouras committed:
pnpm dedupe from rebase

ntsekouras committed:
remove obsolete documentation

anomiex commented on .pnpmfile.cjs:
```diff
@@ -41,6 +41,36 @@ function fixDeps( pkg ) {
+			'@use-gesture/react',
+			'use-memo-one',
+			'uuid',
+			'@wordpress/date',
+			'@wordpress/hooks',
```
Are these last two actually needed?

ntsekouras replied:
I had  copied from the [other PR this list](https://github.com/Automattic/jetpack/pull/41510/files#r1952668188) and while I had checked and updated  the translate functions list, I'll do the same for this one and will remove everything that is not needed!

ntsekouras replied:
I checked the deps and updated them [here](https://github.com/Automattic/jetpack/pull/41602/commits/940f63a87a89c18fb4ef93c9b8b33613fffb351a), where it seems way fewer deps where enough to build successfully. 

I'd really appreciate a final check here.

--edit:

While locally the project was building successfully for me, the CI wasn't happy. By looking at the errors of `E2E: Build jetpack-core` job, it seems that indeed the other deps except the two you commented about, are needed. I updated again [here](https://github.com/Automattic/jetpack/pull/41602/commits/41bd51ec99e68d762c501116c5784925a0d03e10).

anomiex commented on pnpm-lock.yaml:
```diff
@@ -8558,6 +8570,12 @@ packages:
     resolution: {integrity: sha512-SCBrcUttktrqC5facpd29d1bbOee1vWqt+Eg2Wggjvu14A9yWbJKYjfbeHCJYiEJoaFft+WuW4Vq5RSy/QcRjw==}
     engines: {node: '>=18.12.0', npm: '>=8.19.2'}
 
+  '@wordpress/notices@5.17.0':
```
It'd be nice if you could match the versions with the rest of the monorepo to avoid us having both `@wordpress/notices@5.17.0` and `@wordpress/notices@5.19.0` in use.

(although if you don't, we'll wind up having a Renovate PR to do it at some point 🤷)

ntsekouras replied:
Of course. Before merging I'll check the versions of my added packages, because this one got bumped 3 days ago in trunk and missed it.

anomiex commented on projects/packages/forms/src/dashboard/components/layout/index.js:
```diff
@@ -1,26 +1,19 @@
 import { JetpackFooter } from '@automattic/jetpack-components';
+import // __experimentalHStack as HStack, // eslint-disable-line @wordpress/no-unsafe-wp-apis
+'@wordpress/components';
```
This seems odd. You're not actually importing anything from `@wordpress/components` here.

ntsekouras replied:
Wow, that was commented in a really odd way and TBH, I'm not sure I'd see it in the final checks.. Thanks for catching this!

anomiex reviewed (commented):
Gave it a quick skim from a Garage perspective. No major issues, just a few questions.

enejb committed:
Add endpoint example tests

ntsekouras committed:
update deps

ntsekouras committed:
add a PHP test

ntsekouras committed:
update deps again.

ntsekouras committed:
remove some todos

ntsekouras committed:
remove .com style overrides bundle

mcsf commented on projects/packages/forms/src/dashboard/store/selectors.js:
```diff
@@ -0,0 +1,7 @@
+const EMPTY_OBJECT = {};
+const EMPTY_ARRAY = [];
+
+export const getFilters = state => state.filters || EMPTY_OBJECT;
+export const getCurrentQuery = state => state.currentQuery || EMPTY_OBJECT;
+export const getSelectedResponsesFromCurrentDataset = state =>
+	state.selectedResponsesFromCurrentDataset || EMPTY_ARRAY;
```
Nit: is it really necessary to provide the fallbacks (`|| EMPTY`)? Looking at the reducers, they all correctly initialise their state, for example:

( state **= {}**, action ) => { if ( condition ) {} return state }

So I would expect the selectors to retrieve stable objects from `state`. Am I wrong?

ntsekouras replied:
You're right! I'll update.

mcsf commented on projects/packages/forms/src/dashboard/inbox/index.js:
```diff
@@ -1,373 +1,148 @@
+			setSearchParams( previouSearchParams => {
+				const _serachParams = new URLSearchParams( previouSearchParams );
+				_serachParams.set( 'status', newStatusValue );
+				return _serachParams;
```
[serach!](https://en.wikipedia.org/wiki/Sriracha)

```suggestion
				const _searchParams = new URLSearchParams( previouSearchParams );
				_searchParams.set( 'status', newStatusValue );
				return _searchParams;
```

mcsf replied:
eb6f92aef435dd0d89d1332af1a1a894fe96498f :) there was also the "**previou**Search" typo

ntsekouras replied:
Thanks!

mcsf commented on projects/packages/forms/src/dashboard/inbox/export-responses/index.js:
```diff
@@ -0,0 +1,81 @@
+
+import './style.scss';
+
+const ExportResponses = () => {
```
This component (and its directory) might be more aptly named `ExportResponsesButton`.

ntsekouras replied:
Not sure about the directory TBH because it also contains the csv, google buttons etc.. I renamed the component for now though.

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/index.js:
```diff
@@ -0,0 +1,304 @@
+	}
+};
+
+/**
+ * Hook to get the status filter to apply from the URL.
+ * This is the only way to filter the data by `status` as intentionally
+ * we don't want to have a `status` filter in the UI.
+ *
+ * @param {string} urlStatus - The current status from the URL.
+ * @return {string} The status filter to apply.
+ */
+function useStatusFilter( urlStatus ) {
```
Not a hook. Doc should be updated, and func probably renamed to `getStatusFilter`.

ntsekouras replied:
Good catch! It used to be a hook 😄

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/index.js:
```diff
@@ -0,0 +1,304 @@
+const EMPTY_OBJECT = {};
+const MOBILE_BREAKPOINT = 780;
+const getItemId = item => item.id.toString();
+const getPath = response => {
+	try {
+		const url = new URL( response.entry_permalink );
+		return url.pathname;
+	} catch {
+		return '';
+	}
+};
```
Nit: probably deserves a doc block, just like the function that follows it. The argument name `response` may also be a distraction, since the domain-specific term is usually `item`.

ntsekouras replied:
I updated it, but in general I just copied this function here. It was in a `utils` file and it wasn't reused anywhere else.

mcsf reviewed (commented):
I'm still finding my footing in this PR, which is why the comments so far are very local and unimportant (they're impressions I get as I read through the first time). More review to come.

mcsf committed:
Fix typos

edanzer commented:
When testing this, I decided to also check that the Classic view for feedback is still working as expected. I'm seeing some issues there. I double checked on trunk, and these issues do not occur. 
- All my recent form submissions are missing. For some reason, it looks like everything after January 30. 
- The 'From' name has been replaced by an IP address. There's [some logic in class-admin.php](https://github.com/Automattic/jetpack/blob/588036dedca521833af70064bf709455c3ebed48/projects/packages/forms/src/contact-form/class-admin.php#L678) where we use the IP as a fallback if we can't detect an author or emails, so I wonder if we updated something about the data such that we no longer correctly detect the author in this code. 

**Classic feedback on trunk branch**
<img width="1144" alt="classic-feedback-before-dataviews" src="https://github.com/user-attachments/assets/227b3ce9-85d5-4db3-9918-ad4c4f6f951e" />

**Classic feedback on dataviews branch**
<img width="1335" alt="feedback-with-dataviews" src="https://github.com/user-attachments/assets/83530a4b-255e-47af-85fd-e29101c812ee" />

ntsekouras committed:
feedback

ntsekouras commented:
>All my recent form submissions are missing. For some reason, it looks like everything after January 30.
>The 'From' name has been replaced by an IP address

@edanzer I had tested this but didn't paid too much attention to be honest, because for me they seemed okay and I verified that the totals are the same. This seems to be the case for you too probably though, because in both your screenshots you can see that you have `42` items. Can you verify that this is the case and that you don't have differences in `from` as well?


What seems to have changed though is the sort order. In trunk we show `oldest->newest` and in my PR `newest->older`. I'll look into why this happened, although I'd argue it should be like it's here and show the newer responses first..

ntsekouras committed:
change wp-admin list order/orderby

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/index.js:
```diff
@@ -0,0 +1,306 @@
+			if ( field === 'date' ) {
+				const [ year, month ] = value.split( '/' ).map( Number );
+				accumulator.after = new Date( Date.UTC( year, month - 1, 1 ) ).toISOString();
+				accumulator.before = new Date( Date.UTC( year, month, 0 ) ).toISOString();
```
Before I go down the rabbit hole of WP's date queries, let me just ask the question:

`Date.UTC( y, m, 0 )` returns the date at the last day of month `m` at midnight UTC. So there are still 24 hours left in that month. Does the querying logic take this into account, ignoring the time and only checking for the date (i.e. `y-m-d <=` and not `y-m-d h:m <=`)?

ntsekouras replied:
Great catch, thank you!

Querying logic uses the whole timestamp so I fixed it [here](https://github.com/Automattic/jetpack/pull/41602/commits/2ab3762af64f12c64e37e34d200bcde6409bbf72).

I tested by updating in console a `feedback` like this:
`wp.data.dispatch('core').saveEntityRecord('postType', 'feedback', {id: 342, date: '2025-02-28T05:00:00.000Z' }, { throwOnError: true })`

Where the `id:342` should be an existing id for each tester (you can check the network request to get it)... After that you should refresh and apply the `February 2025` filter.

ntsekouras committed:
fix before date filter

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/index.js:
```diff
@@ -0,0 +1,306 @@
+						item.author_name || item.author_email || item.author_url || item.ip
+					);
+				},
+				enableSorting: false,
```
I wonder if users might see value in enabling sorting by sender. But I'll focus on the technical aspects.

ntsekouras replied:
Maybe, but let's iterate on improvements. When this is merged I'll compile a list with enhancement follow ups for sorting, allow multiple filtering values etc..

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/index.js:
```diff
@@ -0,0 +1,306 @@
+			{
+				id: 'date',
+				label: __( 'Date', 'jetpack-forms' ),
+				render: ( { item } ) => dateI18n( 'M j, Y', item.date ),
```
The string format here should be localisable. `M j, Y` isn't very international. :)

ntsekouras replied:
I took this rendering function from the old inbox view. What do you mean though? Use a different date function or change the format?

mcsf replied:
I mean exposing the format to translators:

```js
dateI18n(
  // translators: Date format for ...  ... See https://www.php.net/manual/en/datetime.format.php
  __( 'M j, Y', 'jetpack-forms' ),
  item.date
)
```

although, as I type this, I realise this is dumb: WP already has a site option for date formats, configurable in WP-Admin > General settings. We should use that one.

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/index.js:
```diff
@@ -0,0 +1,306 @@
+				elements: ( filterOptions?.date || [] ).map( _filter => {
+					const date = new Date();
+					date.setDate( 1 );
+					date.setMonth( _filter.month - 1 );
```
Am I correctly assuming that this transformation here is just converting from PHP/WP's 1-indexed values to JS's 0-indexed values? If so, let's make that clear with a comment.

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/index.js:
```diff
@@ -0,0 +1,306 @@
+					date.setDate( 1 );
+					date.setMonth( _filter.month - 1 );
+					return {
+						label: `${ dateI18n( 'F', date ) } ${ _filter.year }`,
```
Hm, what happened here? 😄 

Let's not glue together one `dateI18n` call and a plain parameter. Can't we just set the year on the date and call `dateI18n` with the adjusted format? (Also, we should consider localising this format too.)

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/index.js:
```diff
@@ -0,0 +1,306 @@
+		const recordToShow = data?.find( record => getItemId( record ) === firstValidSelection );
+		if ( ! sidePanelItem && recordToShow ) {
+			setSidePanelItem( recordToShow );
+		} else if ( !! sidePanelItem && ! recordToShow ) {
```
> `if ( !! a )`

Hello, pet peeve. 😄 

Edit: you don't have to do anything here. I just like to point out this kind of syntax, you know me.

ntsekouras replied:
I know you and you know me.. I don't like skipping this if the value is not boolean. Not changing this nit 😄

mcsf commented on projects/packages/forms/src/dashboard/inbox/response.js:
```diff
@@ -6,12 +6,32 @@ import { dateI18n, getSettings as getDateSettings } from '@wordpress/date';
+	return author_name || author_email || ip;
+};
+
+export const formatFieldName = fieldName => {
+	const match = fieldName.match( /^(\d+_)?(.*)/i );
+
+	if ( match ) {
+		return match[ 2 ];
+	}
+
+	return fieldName;
+};
+
+export const formatFieldValue = fieldValue => {
+	if ( isEmpty( fieldValue ) ) {
+		return '-';
+	} else if ( isArray( fieldValue ) ) {
+		return join( fieldValue, ', ' );
+	}
+
+	return fieldValue;
+};
```
Honest question (as in: not a leading question) — is this the best place for this transformation? I had to use the debugger to understand what this was doing. For anyone else: each record sent by the server contains a `fields` object whose keys are numbered:

```js
{
  1_Name: "Alice",
  2_Email: "alice@example.org",
  3_Phone: "1234",
}
```

Anyway, we already manipulate the `fields` property in `InboxView`' (in the `const data = useMemo` call). Why not do this there?

ntsekouras replied:
I just copied this here from the previous utils file because it wasn't reused anywhere else. I'll look at your suggestion though.

mcsf commented on projects/packages/forms/src/dashboard/inbox/response.js:
```diff
@@ -6,12 +6,32 @@ import { dateI18n, getSettings as getDateSettings } from '@wordpress/date';
+export const getDisplayName = response => {
+	const { author_name, author_email, ip } = response;
+	return author_name || author_email || ip;
+};
```
Similarly to the bit below, does it make sense to group this in the `useMemo`?

ntsekouras replied:
Similarly to my above comment. 

>I just copied this here from the previous utils file because it wasn't reused anywhere else. 

This file and the `response` modal/sidepanel item will be re-designed soon and I'd expect to be refactored then. 
I'll look at your suggestion though.

ntsekouras replied:
For now I updated this to match the `render` of `from` field. I think it's better to leave outside of `useMemo` for now and revisit when the IboxResponse  is refactored. It should probably make more sense to keep that value in the REST API if we want to standarize it in general.

mcsf commented on projects/packages/forms/src/dashboard/notices-list.js:
```diff
@@ -0,0 +1,28 @@
+import { store as noticesStore } from '@wordpress/notices';
+import React from 'react';
+
+// Last three notices. Slices from the tail end of the list.
+const MAX_VISIBLE_NOTICES = -3;
```
I'm really picking nits, I know :)

IMO, the constant should be an integer, since that is the semantics of the constant (_how many notices_). Then the `.slice` call should include the negative operator: `.slice( -MAX_VISIBLE_NOTICES )`.

ntsekouras replied:
Copied this part from core :). I'll look at your suggestion though.

mcsf commented on projects/packages/forms/src/dashboard/store/actions.js:
```diff
@@ -0,0 +1,74 @@
+	};
+}
+
+// When we permanently delete some responses, we need to invalidate
+// the filters in the dashboard to reflect the changes.
+export const invalidateFilters = () => {
+	return { type: INVALIDATE_FILTERS };
+};
```
I wonder if this could be improved, but I don't have a suggestion. We can leave it. I really appreciate the comment (and the comment in the Delete action) to make the entanglement of filters and responses clear to other readers.

ntsekouras replied:
Yeah, I don't really like it either, but didn't figure out a better way to do it. I don't think it's that bad in terms with the comments, simple/small store (for now at least) etc..

In general the `shouldInvalidate` function is called when actions from the same store are dispatched, so there was no way to pick the `core` remove items actions. Maybe a registrySelector could do something different, but I don't think so..

mcsf commented on projects/packages/forms/src/contact-form/class-admin.php:
```diff
@@ -85,6 +85,27 @@ public function __construct() {
+	 * @return void
+	 */
+	public function change_feedback_list_order( $query ) {
+		if ( $query->get( 'post_type' ) === 'feedback' ) {
+			$query->set( 'orderby', 'date' );
+			$query->set( 'order', 'desc' );
+		}
```
`pre_get_posts` (the hook to which this is attached) is a pretty delicate hook, and even though this only applies to the Feedback CPT it's still a bit drastic to unconditionally set these query parameters.

Correct me if I'm wrong, but we should at the very least ensure this only happens if the new Inbox interface is going to be rendered; otherwise we are affecting the classic experience unintentionally.

Finally, why not set these on the client in the first place?

ntsekouras replied:
Actually this is just for the classic wp-admin experience that is going to be removed soon, and I have some comments in this function about it.

mcsf commented on projects/packages/forms/src/contact-form/class-admin.php:
```diff
@@ -65,11 +65,11 @@ public function __construct() {
 		add_filter( 'manage_feedback_posts_columns', array( $this, 'grunion_post_type_columns_filter' ) );
 
-		add_action( 'manage_posts_custom_column', array( $this, 'grunion_manage_post_columns' ), 10, 2 );
+		add_action( 'manage_pages_custom_column', array( $this, 'grunion_manage_post_columns' ), 10, 2 );
```
Note for fellow reviewers: it's quite subtle, but AFAICT this change is necessary to accompany the Feedback CPT's switch to `'hierarchical' => true`.

Worthy of a comment, perhaps?

ntsekouras replied:
Maybe, but I think someone didn't know about it and google it, they would find it right away..

I think it's fine though, because really soon the wp-admin list will be removed completely.

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/actions.js:
```diff
@@ -0,0 +1,274 @@
+	id: 'view-response',
+	label: __( 'View response', 'jetpack-forms' ),
+	isPrimary: true,
+	icon: <Icon icon={ seen } />,
```
I found it confusing to see this icon, since it's the same that DataView uses for toggling visibility of fields.

<img width="229" alt="Screenshot 2025-03-07 at 16 50 08" src="https://github.com/user-attachments/assets/3786c06d-2e1f-4051-8c4a-6b8e0e48943b" />

ntsekouras replied:
--cc @jasmussen

ntsekouras replied:
I used the `details` icon, which was also recently added for similar action in site logs.

jasmussen replied:
Yes, this is tricky, in general I'd use a text label here instead of an icon, just "View". But absent the ability to do that, would the envelope icon work? Perhaps "comment content"?
![Screenshot 2025-03-10 at 11 47 11](https://github.com/user-attachments/assets/0354cd7b-e9bc-4506-b791-0edbe8ff2617)

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/actions.js:
```diff
@@ -0,0 +1,274 @@
+		const itemsUpdated = promises.filter( ( { status } ) => status === 'fulfilled' );
+		if ( itemsUpdated.length === items.length ) {
+			// Every request was successful.
+			const numberOfItems = items.length;
```
Necessary?

ntsekouras replied:
You mean the assignment? Or the block of code or something else? If it's about the var, not important but I'm using it 3 times and I have to access the object once. Heavy optimizations 😆 😆

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/actions.js:
```diff
@@ -0,0 +1,274 @@
+							/* translators: The number of responses. */
+							__( '%d responses have been marked as spam.', 'jetpack-forms' ),
+							numberOfItems
+					  );
```
This should be done with `_n`. Same with the errors whenever there is a plurality:

```
_n(
    '%d response has been ...',
    '%d responses have been ...',
    nItems
)
```

and remember that `_n` is [not to be used](https://codex.wordpress.org/I18n_for_WordPress_Developers#Plurals) with `n == 1`.

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/actions.js:
```diff
@@ -0,0 +1,274 @@
+							/* translators: The number of responses. */
+							__( '%d response has been marked as spam.', 'jetpack-forms' ),
+							numberOfItems
+					  )
```
Replace with a plain `__` string, e.g. `__( 'Response marked as spam.' )`

ntsekouras replied:
I think [that's](https://github.com/Automattic/jetpack/pull/41602/commits/75a7991e557a9a9b8ed44cc5ba71be4725433376) what you meant, right?

mcsf replied:
Yes! Thanks.

mcsf commented on projects/packages/forms/src/dashboard/inbox/dataviews/actions.js:
```diff
@@ -0,0 +1,274 @@
+			// There is at least one failure.
+			const numberOfErrors = promises.filter( ( { status } ) => status === 'rejected' ).length;
+			const errorMessage =
+				numberOfErrors === 1
+					? /* translators: The number of responses. */
+					  sprintf( __( 'An error occurred for %d response.', 'jetpack-forms' ), numberOfErrors )
+					: sprintf(
+							/* translators: The number of responses. */
+							__( 'An error occurred for %d responses.', 'jetpack-forms' ),
+							numberOfErrors
+					  );
```
Same notes as for the success message.

mcsf reviewed (commented):
This is looking pretty good, especially for such a big PR.

ntsekouras committed:
feedback

ntsekouras committed:
update `view response` icon

ntsekouras committed:
address i18n feedback

mcsf reviewed (approved):
Thanks for addressing all of my feedback. From a technical standpoint, at least, this is looking good to me!

enejb commented on projects/packages/forms/src/contact-form/class-contact-form-plugin.php:
```diff
@@ -216,6 +216,7 @@ protected function __construct() {
 				'query_var'             => false,
 				'capability_type'       => 'page',
 				'show_in_rest'          => true,
+				'hierarchical'          => true,
```
@ntsekouras Can you explain this change? Why do we need the post type to be hierarchical?

ntsekouras replied:
When `feedback` entities are created the post/source that they come from is saved in `post_parent` attribute of the `feedback`. This feels a bit weird to start with, but since it's been there for a long time, it seemed fine to me to update the post type to `hierarchical` and utilize this field through the REST API for filtering etc..

It doesn't create any UI issues, because the creation of `feedback` entities never happens through a UI (and it shouldn't) and we don't allow the editing of responses in the wp-admin view, which will also be removed soon nevertheless.

ntsekouras commented:
Thank you all for your help!

Possible follow ups can be:

1. Make the new view the [default one](https://github.com/Automattic/jetpack-roadmap/issues/2165). There is already a [PR](https://github.com/Automattic/jetpack/pull/42329) for this.
2. Consider removing the classic wp-admin view altogether
3. Enhancements for filters. We can add support for multiple values selected and we should look into optimizing the current ones and the way they are fetched from the database.
4. Look into adding sorting options.
5. Redesign and refactor the single response view, currently shown in the Modal or as a side panel item depending on the viewport width.
6. p5j4vm-53j-p2#comment-7738 and maybe the UI around which response should be shown when we multi-select items and how.
7. We should look into how to setup e2e tests in a jetpack/package. It seems there isn't a way to do this and needs investigation. Noting that there is some setup for jetpack/plugins.
8. When we do the above we need to write e2e tests for the list and the blocks in general. There are no e2e tests at all right now.
9. See if we can find a way for PHP tests that use a database to be flexible about the tests we can write. There is already a [PR exploring this](https://github.com/Automattic/jetpack/pull/42279).

lezama commented:
> When we do the above we need to write e2e tests for the list and the blocks in general. There are no e2e tests at all right now.

There is at least one Calypso e2e test, I had to silent this one in order to deploy to Simple 😅:

fbhepr%2Skers%2Spnylcfb%2Sgrfg%2Sr2r%2Sfcrpf%2Schoyvfurq%2Qpbagrag%2Ssbezf__fhozvffvbaf.gf%3Se%3Qqsr8s712%26zb%3Q5020%26sv%3Q137%23137-og
